src/ceph.rs

```diff
@@ -934,6 +958,146 @@ impl IoCtx {
+        self.ioctx_guard()?;
+        let obj_name_str = CString::new(object_name)?;
+
+        with_completion(self, |c| unsafe {
+            rados_aio_write(
+                self.ioctx,
+                obj_name_str.as_ptr(),
+                c,
+                buffer.as_ptr() as *const ::libc::c_char,
+                buffer.len(),
+                offset,
+            )
+        })?
+        .await
+    }
+
+    /// Async variant of rados_object_append
+    pub async fn rados_async_object_append(
+        &self,
+        object_name: &str,
+        buffer: &[u8],
+    ) -> RadosResult<i32> {
+        self.ioctx_guard()?;
+        let obj_name_str = CString::new(object_name)?;
+
+        with_completion(self, |c| unsafe {
+            rados_aio_append(
+                self.ioctx,
+                obj_name_str.as_ptr(),
+                c,
+                buffer.as_ptr() as *const ::libc::c_char,
+                buffer.len(),
+            )
+        })?
+        .await
+    }
+
+    /// Async variant of rados_object_write_full
+    pub async fn rados_async_object_write_full(
+        &self,
+        object_name: &str,
+        buffer: &[u8],
+    ) -> RadosResult<i32> {
+        self.ioctx_guard()?;
+        let obj_name_str = CString::new(object_name)?;
+
+        with_completion(self, |c| unsafe {
+            rados_aio_write_full(
+                self.ioctx,
+                obj_name_str.as_ptr(),
+                c,
+                buffer.as_ptr() as *const ::libc::c_char,
+                buffer.len(),
+            )
+        })?
+        .await
+    }
+
+    /// Async variant of rados_object_remove
+    pub async fn rados_async_object_remove(&self, object_name: &str) -> RadosResult<()> {
+        self.ioctx_guard()?;
+        let object_name_str = CString::new(object_name)?;
+
+        with_completion(self, |c| unsafe {
+            rados_aio_remove(self.ioctx, object_name_str.as_ptr() as *const c_char, c)
+        })?
+        .await
+        .map(|_r| ())
+    }
+
+    /// Async variant of rados_object_read
+    pub async fn rados_async_object_read(
+        &self,
+        object_name: &str,
+        fill_buffer: &mut Vec<u8>,
+        read_offset: u64,
+    ) -> RadosResult<i32> {
+        self.ioctx_guard()?;
+        let obj_name_str = CString::new(object_name)?;
+
+        if fill_buffer.capacity() == 0 {
+            fill_buffer.reserve_exact(DEFAULT_READ_BYTES);
+        }
+
+        let result = with_completion(self, |c| unsafe {
+            rados_aio_read(
+                self.ioctx,
+                obj_name_str.as_ptr(),
+                c,
+                fill_buffer.as_mut_ptr() as *mut c_char,
+                fill_buffer.capacity(),
+                read_offset,
+            )
+        })?
+        .await;
+
+        if let Ok(rval) = &result {
+            unsafe {
+                // Completion should have handled any negative (error) values already
+                assert!(*rval >= 0);
```

Comment:
Since the caller is guaranteed to get an unsigned return value, would returning a u32 make more sense here?

Reply:
I originally used i32 in Completion to mirror the blocking interface (which similarly returns i32s even though they must be positive).  But yeah, all the async calls ended up returning an always-positive i32, so let's fix that.